src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1882,10 +1882,99 @@ module T = struct
               match
                 O1trace.sync_thread "validate_transaction_against_staged_ledger"
                   (fun () ->
+                    let get_verification_keys account_ids =
+                      List.fold_until account_ids ~init:Account_id.Map.empty
+                        ~f:(fun acc id ->
+                          let get_vk () =
+                            let open Option.Let_syntax in
+                            let%bind loc =
+                              Transaction_validator.Hashless_ledger
+                              .location_of_account validating_ledger id
+                            in
+                            let%bind account =
+                              Transaction_validator.Hashless_ledger.get
+                                validating_ledger loc
+                            in
+                            let%bind zkapp = account.zkapp in
+                            let%map vk = zkapp.verification_key in
+                            vk.hash
+                          in
+                          match get_vk () with
+                          | Some vk ->
+                              Continue
+                                (Account_id.Map.update acc id ~f:(fun _ -> vk))
+                          | None ->
+                              [%log error]
+                                ~metadata:
+                                  [ ("account_id", Account_id.to_yojson id) ]
+                                "Staged_ledger_diff creation: Verification key \
+                                 not found for party with proof authorization \
+                                 and account_id $account_id" ;
+                              Stop Account_id.Map.empty )
+                        ~finish:Fn.id
+                    in
+                    let valid_proofs () =
```

Comment:
No need to nest this function, it can be extracted out and take `txn` as an argument